### ISSUE TITLE
Planner cooldown, option to give it lower priority.

### ISSUE THREAD

**Issue by DateInJapanese** [PikaMug/Quests#1244]
**What does your feature do?**
It gives the cooldown less priority, making the repeat-cycle override it.
As of now the cooldown MUST be at 0 before being able to possibly start the quest again.
The cooldown starts counting down when the quest is completed.

But for this request I want the cooldown to start down when the quest is completed, but always be set to 0, when the repeat-cycle repeats.
So if there's a repeat cycle of 1 day, and a start-end-date difference of one day. With a cooldown of 5 days. The cooldown would always be set to 0 every day, since the repeat-cycle resets then.

I won't hold you to this statement, but just for reference to what you thought about when first hearing it;
![Screenshot 2020-04-23 at 09 03 01](https://user-images.githubusercontent.com/38226281/80069196-3eba8100-8541-11ea-8776-dbd4b5e8d8be.png)

**How would people use it?**
This would allow you to make actual daily quests, as often requested and/or looked for.
As well as weekly quests etc.
And being able to make those types of quests will allow you to achieve so much more with the plugin.

**Comment by PikaMug** [comment]
Thanks for the reminder. I'll give this some more thought.

**Comment by PikaMug** [comment]
Coming back to this, wouldn't one solution be to set a start date (but no end date), and a cooldown of 24 hours? That way the quest could be taken once daily. No repeat-cycle needed.

**Comment by DateInJapanese** [comment]
No it wouldn't.
As the cooldown starts counting down AFTER the player finished the quest.

What I want is for the player to be able to do the quest every day, becoming available at the same time each day.
Having a 24h cooldown after being finished, could make it like this;
Day 1; Available at 00:00, finish at 23:00.
Day 2; can only start from 23:00 or later, due to the cooldown.

**Comment by PikaMug** [comment]
While I recognize what you're asking isn't currently possible, may I ask what the advantage of doing it that way would be? Do you expect players to crowd a particular quest at midnight each day? Considering most people get at least a few hours of sleep each day, wouldn't a cooldown of say, 20 hours, cover most people that complete the quest same-day?

**Comment by DateInJapanese** [comment]
Well, setting planner dates also sets times for specific timezones.
So midnight for one person might be 9 am for another.
On top of that, the smaller/shorter cooldown option wouldn't work as well-at all, for weekly quests or quests with even longer periods in between.

Some people don't always have time in the morning to play, so they'd finish in the afternoon or late night. But if the weekend comes soon after they might want to play in the morning.

Having a specific time set every day/week that the player can re-play the quest helps with organisation and management.

==

Besides that Quests are often considered a big part of the game content. So being able to provide players with that game content every so often at specific times helps with gathering and keeping them interested and coming back at their chosen time of the day, making it a habit for them to come by and visit. Which is what server owners would want their players to feel and do.

Daily and weekly quests are very often used in all types of games, whether they be RPGs, MMOs, Dating sims, platformers, language practice games, etc. They help the player create a habit and flow.

I also believe that being able to setup quests this way should be a standard option.
Not only because a lot of people ask for how to do this in the discord a lot, but also because I'm willing to bet those who don't ask would be happy to be able to use it.

**Comment by PikaMug** [comment]
@DateInJapanese I've added the setting as described, disabled by default. Tested with a short repeat cycle - it appears to work fine. I'll leave more thorough testing to you.

https://ci.codemc.io/job/PikaMug/job/Quests/137/

**Comment by DateInJapanese** [comment]
Here’s what I noticed happened: Please look at the numbers in the examples below as ‘bigger than’ and ‘smaller than’ rather than as their numeric value.

Ignore cooldown after repeat: true 
==
 [60s] Start-end date difference.
[5s] Repeat-cycle.
[40s] Cooldown.

Finishes Quest (Emphasis on Finishes) >
I get a cooldown of 40s, after which I have to wait 5s 4x (Equaling to 20, for a total of 60s). Before I can take the quest again. When waiting for the 40s cooldown, it displays; you may not take this quest for another Xs. When waiting 5s 4x, it displays the message of; the quest becomes active in Xs. 	*Must say it is a bit confusing when it tells me the quest becomes active in 1 second, but when I try again it tells me it becomes active in 5s. 
==

[60s] Start-end date difference.
[60s] Repeat-cycle.
[5s] Cooldown.

Finishes Quest >
Same as Above. However this one says; Quest will be active in Xs. Xs = the amount of seconds until the next start-end cycle STARTS. This means that it could say: Quest will be active in 40s. Even though the player can already take the quest in 10s, because they finished it 50seconds ago.
 ==

[60s] Start-end date difference.
[5s] Repeat-cycle.
[100s] Cooldown.

Finishes Quest >
I have to wait the 100s cooldown before I get to retake the quest.

==

So in conclusion;
I’d say it doesn’t work yet, since I have to wait the full cooldown, so it doesn’t ignore it;
Even when BOTH the repeat-cycle as well as the start & end date cycle have passed.

What I expected to happen with the Ignore function set to true, would be; When both the Start-End date cycle AND the repeat-cycle have passed, that the quest could be retaken, no matter the cooldown that’s left. 
While instead the start-end cycle acts as a cooldown, just with a different name. (Considering it starts counting only AFTER the quest has been completed.)

**Comment by PikaMug** [comment]
I'm having trouble understanding your explanation, but the gist of it seems to be something isn't working for you regardless of whether the repeat-cycle or cooldown holds a higher value? In my testing (which I repeated just now), both situations work fine for me with a 60s start-end difference. The cooldown is essentially ignored and I'm able to retake the quest once the subsequent repeat-cycle is finished.

Are you taking the quest via command or through an NPC? And were you certain to run _/qa reload_ after editing the quest?

**Comment by DateInJapanese** [comment]
I'm taking mine through an NPC, always do /qa reload.

In the planner I set the "Ignore cooldown after repeat" from false > true.
But it doesn't ignore the cooldown for me.

**Comment by DateInJapanese** [comment]
I see what you mean now.
Yes, they do work when taking them via /quests take, but via the npc it still prioritises the cooldown.

In cases where the repeat-cycle is shorter than the start-end date cycle; it will give out confusing times as shown in the image I send yesterday. Not sure if that's always been that way, but I don't believe it was intended.

**Comment by PikaMug** [comment]
Please give https://ci.codemc.io/job/PikaMug/job/Quests/138/ a try

**Comment by DateInJapanese** [comment]
Running the new build, it doesn't ignore the cooldown on npcs.
But also ran into some other issues regarding understanding how the planner works.

Here is the quest I'm using:
https://pastebin.com/6JwpPnjA
Use block, gravel 2x.
Running 3.9.6-b138

[30s] Start-end difference. (At 0s and 30s mark)
[30s] Repeat
[1d+] Cooldown
Ignore cooldown after repeat = true

Here's what I'm having trouble with understanding:
1. I thought that this setup would make quests available to the player at the 0s and 30s mark.
    However that's not the case.
    I thought it would because that's the time the quest starts (0s), as well as gets repeated (30s).

2. After finishing the quest, it can display the message: "<Quest> will be active in X seconds."
    Where it's possible for `60s > X > 30s`.
    In fact, that's what happened most/all of the time testing.
    This misunderstanding might hook into the first one.

**Comment by PikaMug** [comment]
@DateInJapanese Thanks for catching that. Hopefully https://ci.codemc.io/job/PikaMug/job/Quests/139/ does the trick.

**Comment by DateInJapanese** [comment]
It's going in the right direction.

It still does not work with the NPC, as it still prioritises the Cooldown there.

However, the "<Quest> becomes active in Xs." message, does no longer display times longer than the start-end date difference (-in my case the repeat cycle is shorter than that difference.)
Meaning that with a difference of 30s, I'm able to re-take the quest in ~30s >after< I finish the quest. (So in my case it took the start-end date difference [30s] and started counting from the moment I finished. This also means that it doesn't allow the player to re-take the quest from the 0s and 30s marks.)

So it's not perfect yet, but it's getting closer.

[30s] start-end difference
[5s] repeat
[1d+] cooldown

**Comment by PikaMug** [comment]
@DateInJapanese Glad it's working closer to expectation. For the NPC issue, would you please clarify the following:

- Have you set a Start NPC and GUI item?
- Is `ask-confirmation` true or false in _config.yml_?
- How many quests are attached/linked to the NPC in question?

**Comment by DateInJapanese** [comment]
For the NPC issue:

Yes, I've set a start NPC.
I have no GUI item, it asks me in chat.

`ask-confirmation: true`

1 Quest is linked to the npc.

**Comment by PikaMug** [comment]
@DateInJapanese Thanks for that information. Next up is https://ci.codemc.io/job/PikaMug/job/Quests/140/

**Comment by DateInJapanese** [comment]
Guess it's getting closer again;

Yes, the cooldown is ignored when the ignored setting is set to true.
        and it isn't ignored when set to false, as it should function.

I do feel like there are inconsistencies with the "<Quest> becomes active in Xs." message.
Again;
[30s] duration
[5s] repeat
[1d+] cooldown
ignore: true

After quickly finishing the quest right after accepting it, and trying to take the quest again, the active in Xs message shows up.
It really feels like it sometimes starts counting from 30s down after finishing, while that other times it starts from 25s down.
My console times also show that it sometimes counts from 25, while other times from 30.

Not that it should be doing that in the first place, as again, in this case it should start from the 0s mark and 30s mark, where the player can re-take the quest. But it's progress nonetheless.

**Comment by Starkitobonito** [comment]
I want the same as DateInJapanese. I already downloaded the latest developer version, and im accepting the quest from a NPC with GUI. This is the planner:

(THIS IS A TEST)
    planner:
      start: 31:6:2020:8:46:0:America/Chicago
      end: 31:6:2020:8:46:30:America/Chicago
      repeat: 30
      cooldown: 86400
      override: true

So, if is the ignore cooldown is active, why the quest dont restart when finish the repeat time? if is in 30s of time.
 Im looking make quests that restart every day at 00:00

**Comment by PikaMug** [comment]
@DateInJapanese @Starkitobonito At this point, I'm going to call that delay of a few seconds an acceptable margin of error.  If either of you start creating quests with longer time intervals and see a delay of more than 5 seconds, let me know.

**Comment by Starkitobonito** [comment]
How should I configure this to restart the mission at 00:00 every day?

    planner:
      start: 28:6:2020:1:0:0:America/Chicago
      end: 28:6:2020:23:59:59:America/Chicago
      repeat: 72000
      cooldown: 72000
      override: true

**Comment by PikaMug** [comment]
@Starkitobonito Just change start time to midnight (`0:0:0` instead of `1:0:0`) and your `repeat` to `86400`

**Comment by Starkitobonito** [comment]
I set this planner and when i finish the quest, the cooldown is for 2 days. Why?

pick: https://imgur.com/lHxz4bj

    planner:
      start: 28:6:2020:0:0:0:America/Chicago
      end: 28:6:2020:23:59:59:America/Chicago
      repeat: 86400
      cooldown: 86400
      override: true

**Comment by PikaMug** [comment]
@Starkitobonito Which build number from Jenkins are you using? Also, it's talking about the repeat delay, not cooldown.

**Comment by Starkitobonito** [comment]
git-paper-93 (MC:1.16.1)

**Comment by PikaMug** [comment]
@Starkitobonito No, your Quests version. From https://ci.codemc.io/job/PikaMug/job/Quests/

**Comment by Starkitobonito** [comment]
quests 3.9.6 - b138

**Comment by Starkitobonito** [comment]
Yes, i set the repeat for 1 day, but when i finish the quests, tell me that left 1 day and 19 hours (i finished the quest at 5:00 am)

planner:
  start: 28:6:2020:0:0:0:America/Chicago
  end: 28:6:2020:23:59:59:America/Chicago
  repeat: 86400
  cooldown: 86400
  override: true

**Comment by PikaMug** [comment]
@Starkitobonito There's the problem. Use build 140.

**Comment by Starkitobonito** [comment]
Its working now, ty

**Comment by DateInJapanese** [comment]
@Starkitobonito Do your quests now become available again at midnight every day? No matter when the quest is completed?
Because it doesn't work like that for me yet, and I'm also using b140.

Not sure if I was clear about that @PikaMug.
But it was mentioned at the bottom of my previous comment.

**Comment by Starkitobonito** [comment]
Yes @DateInJapanese , now every day at midnight the quests can get taked again for all the players. Works like a real daily quest.

**Comment by DateInJapanese** [comment]
Mhh alright, guess I'll try it again then.
Thanks for the response.

**Comment by DateInJapanese** [comment]
I see now, works for me as well.
Not sure what I did wrong last time, but what was different last time was that;
My `repeat != cooldown != duration != repeat.`

But now I've got it setup to where they're all just as long, and it works ofc.
